### DISCUSSION TITLE
How to activate bodies from a step listener?

### DISCUSSION THREAD

**Discussion by hhyyrylainen** [jrouwe/JoltPhysics#784] [Q&A] [answered]
I'm using a step listener to implement game object movement. I've been seeing a bug where when an object is initially touched by something else it can shoot off at high velocity. Just today I had the thought that that is probably caused by the bodies being asleep while accumulating velocity from repeated impulse applying. But I ran into a major issue: I apply the impulse with a step listener and I can't figure out how to activate the body without this assert triggering: `JPH_ASSERT(body.IsInBroadPhase());`

This is the code I tried:
```c++
physicsSystem->GetBodyInterfaceNoLock().ActivateBody(bodyId);
```

I've already been using the no-lock interface for setting the body velocity and angular velocity, which seems to work fine. But that above code causes an assert to trigger within like a second of things starting to happen.

I tried the locking variant but that just permanently locks up the entire process:

```c++
physicsSystem->GetBodyInterface().ActivateBody(bodyId);
```

Is there a correct way to activate bodies from a step listener that I missed?

**Comment by jrouwe** [comment, marked as answer]
The assert ```JPH_ASSERT(body.IsInBroadPhase());``` means that you're trying to activate a body that hasn't been added to the ```PhysicsSystem``` (using ```BodyInterface::AddBody```).

If you want a body to wake up when applying an impulse you need to call ```BodyInterface::AddImpulse``` instead of ```Body::AddImpulse```.  Otherwise you can indeed start accumulating velocity on a sleeping body (and if the body hasn't been added to the world it's the same thing). I'll have to add an assert there.

**Reply by hhyyrylainen** [reply]
Ah, turns out I have an underlying bug with bodies that are not in the broadphase being in the list of stuff to apply movement control to. I initially thought when I saw that assertion that I was doing something unexpected (as I had previously seen that same assert when I accidentally turned on multithreading for body adding to world despite using the basic single body add API). But now quickly testing adding a `if (!body.IsInBroadPhase()) return;` to my method, that stops the crashing when using that no lock interface. So yeah I think I did initially find the right API, I just had a bug elsewhere.

Thank you for the response. I'll investigate what's the root cause for bodies not being in the broadphase in my setup.

**Reply by hhyyrylainen** [reply]
Yeah that was exactly it, I had a bug in my code. There was a one frame window when part of my code thought that a body was already added to the world, but the code for adding the body to the world had not ran yet. I fixed that by skipping the physics body control code from running if the body wasn't actually added to the world yet.

So yeah, if anyone ever stumbles upon this discussion `physicsSystem->GetBodyInterfaceNoLock().ActivateBody(bodyId);` is the right API to use and it works perfectly fine as long as there isn't a bug elsewhere with body adding to the world happening too late.
